### PR TITLE
Refactor SnippetRegistry with SnippetGenerator types

### PR DESCRIPTION
**Pull request by amckinney**
Fixes https://github.com/fern-api/fern-python/issues/368

This preserves the same snippet generation behavior, but makes a few significant changes to the implementation, including:

* The `TypeDeclarationSnippetGenerator` produces snippets for every type, including aliases, enums, objects, and unions.
* The new `SnippetWriter` implements a variety of snippet generation helper functions and is used to write snippets.
* There is a recursive relationship between the `TypeDeclarationSnippetGenerator` and the `SnippetWriter` (by necessity).
* The `SnippetRegistry` is still used, but only as as an object that stores all the snippets (to be written in the `snippet.json`). All other functionality is moved to the `SnippetWriter`.

Nearly all of the snippet-specific abstractions are defined in the `fern_python.snippet` module, and a few structural changes are introduced to prevent import cycles including:
* The pydantic model context and core utilities are hoisted up a layer to prevent an import cycle in the model generator (because the snippet package requires access to the base context).
* The `SnippetWriter` constructor helper is defined in the `fern_python.utils` package, and copied into the pydantic model generator package to prevent another import cycle.

In general, the contract is that a type generator will produce a snippet using the `SnippetWriter`, which will then be registered in the `SnippetRegistry`. After all generation is complete, the `SnippetRegistry` is serialized as a `snippet.json`.

### CODE REVIEW

**Comment by amckinney** [comment]
@dsinghvi To prevent an import cycle between the `fern_python.snippet` module and the `fern_python.generators.pydantic_model` module, the `context` and `core_utilities` functions were hoisted above so that we have the following directory structure:

```
generators
├── context
├── core_utilities
├── fastapi
├── pydantic_model
└── sdk
```

I kept the naming of the types in `context` and `core_utilities` the same (e.g. `PydanticGeneratorContext`) since the structure actually retains the same sematnics - we treat the model contexts as the "base" context used by both the SDK and FastAPI contexts, so placing it up a layer isn't so bad.

I'm open to other names (e.g. `BaseContext`, `ModelContext`, etc), but left it as-is for now to keep the changes to a minimum.

File: src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py [hunk 1]
```diff
@@ -131,9 +131,7 @@ def get_class_reference_for_type_name(self, type_name: ir_types.DeclaredTypeName
     def _must_import_after_current_declaration(self, type_name: ir_types.DeclaredTypeName) -> bool:
         if self._type_name is None:
             return False
-        is_circular_reference = HashableDeclaredTypeName.of(self._type_name) in self._context.get_referenced_types(
-            type_name
-        )
+        is_circular_reference = self._context.do_types_reference_each_other(self._type_name, type_name)
```
Comment (amckinney) [inline src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py:134]: @dsinghvi Here's the change required to get the update_forward_refs functionality working again. This function was already defined on the context, so it's definitely what we should be using here.

File: src/fern_python/snippet/snippet_writer.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    def get_module_path_for_declared_type_name(
+        self,
+        name: ir_types.DeclaredTypeName,
+    ) -> AST.ModulePath:
+        module_path = tuple([directory.snake_case.unsafe_name for directory in name.fern_filepath.package_path])
+        if len(module_path) > 0:
+            # If the type is defined in a subpackge, it needs to be imported with the 'resources'
+            # intermediary key. Otherwise the types can be imported from the root package.
+            module_path = ("resources",) + module_path
```
Comment (dsinghvi) [inline src/fern_python/snippet/snippet_writer.py:58]: looks like this is a place where we can encounter a bug -- ideally we should be reusing the logic in the type declaration referencers when constructing the snippet

Comment (dsinghvi) [reply src/fern_python/snippet/snippet_writer.py:58]: might be a lot to change now but can we file an issue for this

File: src/fern_python/snippet/snippet_writer.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            qualified_name_excluding_import=(),
+            import_=AST.ReferenceImport(
+                module=AST.Module.snippet(
+                    module_path=self.get_module_path_for_declared_type_name(
+                        name=name,
+                    ),
+                ),
+                named_import=name.name.pascal_case.unsafe_name,
+            ),
+        )
```
Comment (dsinghvi) [inline src/fern_python/snippet/snippet_writer.py:48]: same comment as below -- we should be reusing the logic in the type declaration referencer

File: src/fern_python/snippet/snippet_registry.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        if method is ir_types.HttpMethod.GET:
+            return generator_exec.EndpointMethod.GET
+        if method is ir_types.HttpMethod.POST:
+            return generator_exec.EndpointMethod.POST
+        if method is ir_types.HttpMethod.PUT:
+            return generator_exec.EndpointMethod.PUT
+        if method is ir_types.HttpMethod.PATCH:
+            return generator_exec.EndpointMethod.PATCH
+        if method is ir_types.HttpMethod.DELETE:
+            return generator_exec.EndpointMethod.DELETE
```
[the remark this answers is not in the document]
Comment (amckinney) [reply src/fern_python/snippet/snippet_registry.py:123]: Great call, TIL `assert_never` in `typing_extension` - have only ever seen it in our TypeScript repos thus far.

File: src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py (hunk 1 above)
Comment (dsinghvi) [reply src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py:134]: weird the context does the same thing: 

```python
def do_types_reference_each_other(self, a: ir_types.DeclaredTypeName, b: ir_types.DeclaredTypeName) -> bool:
        return self.does_type_reference_other_type(a, b) and self.does_type_reference_other_type(b, a)

    def does_type_reference_other_type(
        self, type: ir_types.DeclaredTypeName, other_type: ir_types.DeclaredTypeName
    ) -> bool:
        return HashableDeclaredTypeName.of(other_type) in self.get_referenced_types(type)
```

Comment (dsinghvi) [reply src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py:134]: we'll see if someone shouts

**Review by dsinghvi — APPROVED** [review]

File: src/fern_python/generators/sdk/client_generator/endpoint_function_generator.py [hunk 5]
```diff
@@ -769,6 +738,130 @@ def _get_literal_header_value(self, header: ir_types.HttpHeader) -> Optional[str
...
+        return AST.Expression(
+            AST.FunctionInvocation(
+                function_definition=AST.Reference(
+                    qualified_name_excluding_import=(get_endpoint_name(self.endpoint),),
+                ),
+                args=args,
+            ),
+        )
+
+    def _get_snippet_for_inlined_request_body_properties(
```
[the remark this answers is not in the document]
Comment (amckinney) [reply src/fern_python/generators/sdk/client_generator/endpoint_function_generator.py:817]: Unless I'm mistaken, looks like we don't support that today - we should cut an issue to support examples for both file upload and bytes requests. Today we have:

```yaml
  HttpRequestBody:
    union:
      inlinedRequestBody: InlinedRequestBody
      reference: HttpRequestBodyReference
      fileUpload: FileUploadRequest
      bytes: BytesRequest

  ExampleRequestBody:
    union:
      inlinedRequestBody: ExampleInlinedRequestBody
      reference: types.ExampleTypeReference
```

File: src/fern_python/snippet/snippet_writer.py (hunk 2 above)
Comment (amckinney) [reply src/fern_python/snippet/snippet_writer.py:58]: Issue here: https://github.com/fern-api/fern-python/issues/441

File: src/fern_python/snippet/snippet_writer.py (hunk 3 above)
Comment (amckinney) [reply src/fern_python/snippet/snippet_writer.py:48]: Issue here: https://github.com/fern-api/fern-python/issues/441

File: src/fern_python/generators/sdk/client_generator/endpoint_function_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amckinney) [reply src/fern_python/generators/sdk/client_generator/endpoint_function_generator.py:817]: Issue here: https://github.com/fern-api/fern/issues/2439

**Pull request merged**